### PR TITLE
73 climate matching function

### LINKED ISSUE #73: create function for climate matching
@damianooldoni within the RIPARIAS project @SanderDevisscher and me developed a procedure for climate matching, the idea is described in [this issue](https://github.com/inbo/easin/issues/39) and further developed in the RIPARIAS repo for crayfish and aquatic plants.

This issue is meant to make this into a function within the [trias](https://github.com/trias-project) package.

### LINKED ISSUE #74: Expand climate matching function with single species maps
The idea is to expand the function from #73 with single_species_maps as a new item in the functions return list. 
These maps should be made as a sublist of taxonkeys containing a leaflet with all scenarios as basegroups.

so the final output of this expansion should be callable as `ouput$single_species_maps$taxonkey_1`

### PR DESCRIPTION
**Pull request by SanderDevisscher**
Fixes #73 
Fixes #74 

I've created a new function for the Trias package which uses GBIF observations to create a series of rudimentary climate matching outputs. 

To use this function 3 datapackages (observed, future & legend)  have been added. These contain the shapes for the observed and future climate zones based on the Koppen Geiger classifications and their respective legends. These shapes originate from Rubel & Kottek 2010 and Beck et al 2018. 

The flowchart below depicts the steps taken to create the outputs:

CM_function_flow

As extra functionality 3 extra outputs were added, namely a current climate map, future climate maps and single species maps

### CODE REVIEW

**Comment by damianooldoni** [comment]
Hi @SanderDevisscher: I see you are still pushing commits so I converted the PR to draft. Please, convert it back to PR if you are done and you think it's ready for review. If you think you need some intermediate checks at this stage from my side, I can do it in draft mode as well.

File: R/climate_match.R [hunk 1]
```diff
@@ -653,7 +653,7 @@ climate_match <- function(region,
     }  
   }else{
     current_climate_map <- "maps were disabled"
```
Comment (damianooldoni) [inline R/climate_match.R:655]: @SanderDevisscher: if maps are disabled, this three slots should be set to `NULL`. I like you want to inform the user about this. Let's do it using `message("maps are disabled")`.

Comment (SanderDevisscher) [reply R/climate_match.R:655]: In the current version the map slots are filled with the "maps are disabled" string. 
Do you want me to add an additional warning message ?

Comment (SanderDevisscher) [reply R/climate_match.R:655]: @damianooldoni ok

Comment (SanderDevisscher) [reply R/climate_match.R]: done

**Comment by SanderDevisscher** [comment]
@damianooldoni are there still issues with this PR ?

**Comment by damianooldoni** [comment]
Can you try to solve all problems found by running a check (via RStudio button in Build pane or `devtools::check(document = FALSE)`)? For any problem, let me know! Then, I will try to find time for a real code review and writing tests.

**Comment by SanderDevisscher** [comment]
@damianooldoni besides some notes which I think I can fix myself, I also get this warning:

```
checking LazyData ... WARNING
    LazyData DB of 5.1 MB without LazyDataCompression set
    See �1.1.6 of 'Writing R Extensions'
```

What to do with this ?

**Comment by damianooldoni** [comment]
I checked the suggested [documentation](https://cran.r-project.org/doc/manuals/R-exts.html#Data-in-packages) and I added `LazyDataCompression: bzip2` to DESCRIPTION. Let's see if it works.

**Comment by SanderDevisscher** [comment]
@damianooldoni should I fix all notes ?

**Comment by SanderDevisscher** [comment]
@damianooldoni all warnings and most of the notes have been fixed

**Comment by damianooldoni** [comment]
Let's see what the automatic checks say! If it's green, I will think I can start the review/writing tests.

**Comment by SanderDevisscher** [comment]
There seems to be something wrong with test-update_download_list.R:
```
== Failed tests ================================================================
-- Error (test-update_download_list.R:25:3): gbif download added correctly to list GBIF downloads and other status updated --
Error: Error: 'arg' must be NULL or a character vector
Backtrace:
    x
 1. \-trias::update_download_list(...) test-update_download_list.R:25:2
 2.   \-readr::write_tsv(...)
 3.     \-readr::write_delim(...)
 4.       \-vroom::vroom_write(...)
 5.         \-base::match.arg(escape)

[ FAIL 1 | WARN 1 | SKIP 0 | PASS 811 ]
Error: Error: Test failures
Execution halted

1 error x | 0 warnings v | 2 notes x
Error: Process completed with exit code 1.
```

**Comment by damianooldoni** [comment]
Thanks. I will give a look to this test and I will review/test your function. Thanks a lot for improving the functionalities of `trias` package!

**Comment by damianooldoni** [comment]
Not easy to understand what was going wrong. But at the end I got it: it's a problem with Windows and new major version release of `readr`. The data are read lazily by default and you cannot modify them. From `read_*` function documentation:

> Note, lazy reading on windows will lock the file until all the data has been read from it, if you run into this issue set lazy = FALSE.

With new readr 2.x arg `quote_escape` is also deprecated and substituted by `escape`. Warning returned otherwise.

**Comment by SanderDevisscher** [comment]
- [ ] add example to documentation
- [ ] rerun defaults to "no"

**Comment by SanderDevisscher** [comment]
Apparently the region layer from rworldmap does also include the overseas territories of European countries (& the entirety of Russia). This is not as I intended the climate matching. Europe as example below.

```
library(leaflet)
library(rworldmap)

worldmap <- getMap(resolution = "low")

europe <- subset(worldmap, tolower(worldmap$Stern) == "europe")

leaflet(europe) %>% 
  addTiles() %>% 
  addPolylines()
```
image

Possible solutions I see are:

1. add new "correct" shapes to the datapackage
2. make climate matching global => remove region parameter & logic altogether
3. use another service like rworldmap => which ?
4. add logic to correct rworldmap faults ea overwrite incorrect regions
5. do nothing since these overseas territories are part of Europe 

@timadriaens @damianooldoni what do you think?

**Comment by damianooldoni** [comment]
@SanderDevisscher: this is how I see this problem.

First, saying that rworldmap is incorrect and we have to correct its faults as mentioned at (4) is incorrect! `Rworldmap` is perfectly correct. It's just that you are seeing Europe physically, i.e. as a continent, while `Rworldmap` conceives Europe politically so including all States included in the continent PLUS their oversea territories. 

Returning to the question, if you really want to limit the climate match to the continental Europe, an easy way is to hardcode the values of `europe$ADMIN` you need, or other way around, the values you want to exclude, e.g. `"Greenland"`, `"Israel"`, `"West Bank"`, `"Sint Maarten"`. Then, to make this function reusable for everybody in the world and future-we(!), I would suggest to write the climate function as general as possible, i.e. allowing to pass a sp object (or sf). And in this way, if we save the filtered data folder `data` as .rda file, we can pass `europe` to the function smoothly. If somebody else wants to apply the function for another continent, then just need to pass another sp (sf?) object. Shortly, **option 1** 😄 

But, at the same time, I don't know how incorrect and nonsense is to apply the climate match function to overseas territories. I leave this to @timadriaens and you.

**Comment by SanderDevisscher** [comment]
@damianooldoni I forgot the quotes around "faults" :-/

You make some good points about re-usability and I'll try to rewrite the function so it allows the use of a SP (or sf) object as region. While maintaining the use of rworldmap as it is. 

something like:

```
if(is.character(region)){
region_shape <- current region logic (ea get shape from rworldmap)
}else{
region_shape <- trycatch({
get(region), 
error=function(cond) {
message(cond)
return(NA)
},
warning=function(cond) {
            message(cond)
            return(NA)
})

if(class(region_shape) == sf){
region_shape <- as(region_shape, "spatial")
}

if(class(region_shape) %in% c(sp classes of intrest)){
}else{
stop("region is not a spatial object of the correct class")
}

region_shape <- sptransform(region_shape, crs_wgs)
```

**Comment by SanderDevisscher** [comment]
@damianooldoni I've added an example, switched the default state of the rerun to "no" and allowed the use of a custom perimeter shape. Besides the warnings & errors from the check below I think (and hope) the function is now ready for your review.
```
== Warnings ====================================================================
-- Warning (test-output_gbif_has_distribution.R:4:3): gbif_has_distribution with user parameters --
`distinct_()` was deprecated in dplyr 0.7.0.
Please use `distinct()` instead.
See vignette('programming') for more help
This warning is displayed once every 8 hours.
Call `lifecycle::last_warnings()` to see where this warning was generated.
Backtrace:
  1. testthat::expect_true(gbif_has_distribution(140563025, country = "BE")) test-output_gbif_has_distribution.R:4:2
 16. dplyr::distinct_(., .dots = names(user_properties))
 17. dplyr:::lazy_deprec("distinct")

== Failed tests ================================================================
-- Error (test-update_download_list.R:17:3): gbif download added correctly to list GBIF downloads and other status updated --
Error: unused argument (lazy = FALSE)

[ FAIL 1 | WARN 1 | SKIP 0 | PASS 811 ]
Error: Test failures
Execution halted

1 error x | 1 warning x | 3 notes x
Error: R CMD check found ERRORs
Execution halted

Exited with status 1.
```

**Review by damianooldoni — COMMENTED** [review]
Sorry @SanderDevisscher for this late answer. I have just solved the small bug behind the automatic test failures. I was also checking your function, but I stopped: I was stuck as I could not find anymore which curly bracket I erroneously deleted etc. 

You can merge this PR if you need to, but these points should be tackled sooner or later:
1. GBIF instead of gbif in documentation
2. `taxon_key` instead of `taxonkey`, `zip_file` instead of `zipfile` etc.: ALL functions in this package use snake_case. I don't find any reason to have an exception for this function.
3. This function is way too long. Splitting in subfunctions will help readibility and debugging: future-you, the maintainer (that's me 😮 ) and anyone else trying to understand the workflow will be grateful. Notice that these subfunctions can be written in the same R file, just append them below the "main" function. For example, the leaflet part is clearly something you can put in a subfunction, isn't?
4. Unit-testing are needed: I can do it, but first 3 needs to be solved.
5. About the entire workflow behind this function, I really suggest you to write a vignettte about. I mean, it's a pity the schema you drew in https://github.com/trias-project/trias/pull/77#issue-958159892 would be "lost" in this PR.

So, if you need, I approve this PR: this is what you prefer, I read your comments correctly. I will then translate all my thoughts in some issues. Otherwise, I will officially "request changes" and you can work further on this branch. Let me know!

Anyway, I appreciate really a lot your work and what this function does! 💪  A great function, an added value for this package. It's now a question of refactoring and polishing.

**Comment by SanderDevisscher** [comment]
> Sorry @SanderDevisscher for this late answer. I have just solved the small bug behind the automatic test failures. I was also checking your function, but I stopped: I was stuck as I could not find anymore which curly bracket I erroneously deleted etc.

Just glad we can make some progress 
 
> You can merge this PR if you need to, but these points should be tackled sooner or later:

No specific need. But letting the branches diverge has proven a recipe for conflicts. Also since it is very busy atm I think its best to merge this branch after I tackled remarks 1 & 2 and to work with issues for the rest.  

> 1. GBIF instead of gbif in documentation
> 2. `taxon_key` instead of `taxonkey`, `zip_file` instead of `zipfile` etc.: ALL functions in this package use snake_case. I don't find any reason to have an exception for this function.

I'll fix the issues above before the merge. Those below should be turned into issues. 

> 3. This function is way too long. Splitting in subfunctions will help readibility and debugging: future-you, the maintainer (that's me 😮 ) and anyone else trying to understand the workflow will be grateful. Notice that these subfunctions can be written in the same R file, just append them below the "main" function. For example, the leaflet part is clearly something you can put in a subfunction, isn't?

I agree I just don't exactly know how besides this is a huge work for which I do not have time atm.  

> 4. Unit-testing are needed: I can do it, but first 3 needs to be solved.

I'll do this if you show me how (so I can do it for future functions I write)

> 5. About the entire workflow behind this function, I really suggest you to write a vignettte about. I mean, it's a pity the schema you drew in [#77 (comment)](https://github.com/trias-project/trias/pull/77#issue-958159892) would be "lost" in this PR.

Again I agree but do not know how to make a vignette

> So, if you need, I approve this PR: this is what you prefer, if I understood your comment above correctly. I will then translate all my thoughts in some issues. Otherwise, I will officially "request changes" and you can work further on this branch. Let me know!

> Anyway, I really appreciate your work and what this function does! 💪 A great function, an added value for this package. It's now a question of refactoring and polishing.

**Comment by damianooldoni** [comment]
Thanks @SanderDevisscher. I agree 100% with your answer. I will be very glad to provide you with the help you need for both unit tests and vignettes. Maybe a kind of "sprint" day in a remote place without being by nobody and any other request, will be the most effective way for this.

About snake case, be careful: not easy to change all of them without making an error. I speak about my personal experience. But you know your code better than me, of course, and so, it will work. I will now bump the version if still not done. So, please, pull from remote within 5 minutes.

**Comment by SanderDevisscher** [comment]
@damianooldoni I fixed issues 1 & 2 (partially, I'll implement more snake_case when I split the function). 
Can you approve the PR and merge ?

**Pull request merged**